### PR TITLE
DDP-4053 support permanent announcements

### PR DESCRIPTION
**Pull request by yufengwng**
Original PR: #19.

Since there was significant changes to events, the merge was a bit ugly so I created a new branch and new PR. See the old PR for comments and notes.

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/UserGovernanceDao.java [hunk 1]
```diff
@@ -112,6 +112,19 @@ default int deleteAllGovernancesForProxy(long proxyUserId) {
         return findGovernancesByParticipantAndStudyGuids(participantGuid, studyGuid).filter(Governance::isActive);
     }
 
+    @UseStringTemplateSqlLocator
+    @SqlQuery("queryGovernancesByParticipantAndStudyIds")
+    @RegisterConstructorMapper(Governance.class)
+    @RegisterConstructorMapper(GrantedStudy.class)
+    @UseRowReducer(GovernanceWithStudiesReducer.class)
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/UserGovernanceDao.java:119]: Some really nice jdbi work here--filling out `grantedStudies` inside `Governance` on-the-fly.  Along with the `select_all_governances_with_granted_studies` bits in the `stg` file, this is really slick.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/EventConfigurationDto.java [hunk 2]
```diff
@@ -95,7 +97,9 @@ public EventConfigurationDto(long eventConfigurationId,
                                  Long workflowStateId,
                                  Boolean triggerAutomatically,
                                  String dsmNotificationEventType,
-                                 Long messageTemplateId,
+                                 Long announcementMsgTemplateId,
+                                 Boolean announcementIsPermanent,
+                                 Boolean announcementCreateForProxies,
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/EventConfigurationDto.java:102]: How about `createAnnouncementForProxies`?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRoute.java [hunk 3]
```diff
@@ -57,46 +56,48 @@ public GetUserAnnouncementsRoute(I18nContentRenderer renderer) {
...
                     throw ResponseUtil.haltError(response, HttpStatus.SC_INTERNAL_SERVER_ERROR, err);
                 }
 
-                // Currently, announcements read are automatically deleted.
-                int numDeleted = announcementDao.deleteByIds(announcementIds);
-                if (numDeleted != announcements.size()) {
+                Set<Long> transientAnnouncementIds = announcements.stream()
+                        .filter(ann -> !ann.isPermanent())
+                        .map(UserAnnouncement::getId)
+                        .collect(Collectors.toSet());
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRoute.java:89]: Good call on `Set` here to protect against "no such row" if something goes haywire server-side and returns redundant rows.

File: pepper-apis/src/main/resources/db-changes/schema/DDP-4053-permanent-announcements.xml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    <changeSet author="yufeng" id="20191217-permanent-user-announcements">
+        <renameColumn tableName="user_announcement" oldColumnName="participant_user_id" newColumnName="user_id" columnDataType="bigint"/>
+
+        <addColumn tableName="user_announcement">
+            <column name="guid" type="varchar(36)"/>
+            <column name="is_permanent" type="boolean"/>
+        </addColumn>
+
+        <update tableName="user_announcement">
+            <column name="guid" valueComputed="uuid()"/>
```
Comment (zyme) [inline pepper-apis/src/main/resources/db-changes/schema/DDP-4053-permanent-announcements.xml:33]: We should have a unique constraint here, right?  I'm okay with `uuid()` as backfill.  Thanks for using `DBUtils.uniqueUUID4()` when creating rows in the normal flow.

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java [hunk 5]
```diff
@@ -177,4 +189,74 @@ public void test_basicContentStyle() {
...
+            assertEquals(1, found.size());
+            assertEquals(guid, found.get(0).getGuid());
+        });
+
+        given().auth().oauth2(token)
+                .pathParam("userGuid", testData.getUserGuid())
+                .pathParam("studyGuid", testData.getStudyGuid())
+                .when().get(url)
+                .then().assertThat()
+                .statusCode(200).contentType(ContentType.JSON)
```
Comment (zyme) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java]: There's a fair amount of copy-pasta here--can you factor this out to a private method?

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java [hunk 6]
```diff
@@ -177,4 +189,74 @@ public void test_basicContentStyle() {
...
+                .body("$.size()", equalTo(1))
+                .body("[0].guid", not(isEmptyOrNullString()))
+                .body("[0].permanent", equalTo(true))
+                .body("[0].message", equalTo(msgTemplate.getTemplateText()));
+
+        TransactionWrapper.useTxn(handle -> {
+            handle.attach(UserAnnouncementDao.class).deleteAllForUserAndStudy(testData.getUserId(), study.getId());
```
Comment (zyme) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java:227]: I think `deleteAllForUserAndStudy()` is covered by the `@After`?

**Review by zyme — COMMENTED** [review]
Looks good overall.  One issue when running tests locally was a checkstyle flag for import of `Ignore` in `EventServiceTest`.

**Review by zyme — DISMISSED** [review]

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java (hunk 6 above)
Comment (yufengwng) [reply pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java:227]: In this test case, we created a new study which is different from the one `@After` looks at, so we're calling delete for this one.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/EventConfigurationDto.java (hunk 2 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/EventConfigurationDto.java:102]: Renamed

File: pepper-apis/src/main/resources/db-changes/schema/DDP-4053-permanent-announcements.xml (hunk 4 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/db-changes/schema/DDP-4053-permanent-announcements.xml:33]: Ah yes, I missed it! I have added the UK to the column definition above.

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java (hunk 6 above)
Comment (zyme) [reply pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetUserAnnouncementsRouteTest.java:227]: Got it!

**Review by zyme — APPROVED** [review]

**Comment by zyme** [comment]
please run tests

**Pull request merged**